src/System.Private.CoreLib/shared/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -532,38 +531,37 @@ public static void Prelink(MethodInfo m)
 
         public static void PrelinkAll(Type c)
         {
-            if (c == null)
+            if (c is null)
             {
                 throw new ArgumentNullException(nameof(c));
             }
 
             MethodInfo[] mi = c.GetMethods();
-            if (mi != null)
+
```

Comment:
This is ok because no override of GetMethods will return null?

Reply:
Right, they shouldn't. The documentation for the API is:
> An array of MethodInfo objects representing all methods defined for the current Type that match the specified binding constraints.
>
> -or-
>
> An empty array of type MethodInfo, if no methods are defined for the current Type, or if none of the defined methods match the binding constraints.

This should also match the concurrent PR against Type.

Reply:
Hmm.  @jkotas, any opinion on this?

Even if the docs say you shouldn't do this, and even if our annotations will say it doesn't return null, and even though this is super rare, since we've always previously checked and since it is technically possible for someone to override this and return null, and since this doesn't seem like a particularly hot path we need to eliminate branches on, I'd be inclined to leave the null check in place.

Reply:
I am fine with deleting the null check. Nobody will notice.

This is useless API with exactly zero use: https://apisof.net/catalog/System.Runtime.InteropServices.Marshal.PrelinkAll(Type) . The only reason why it exists in .NET Core is a check box for NS2.0.

Reply:
> This is useless API with exactly zero use … The only reason why it exists in .NET Core is a check box for NS2.0

The gift that keeps on giving :)

> I am fine with deleting the null check. Nobody will notice.

Ok, sounds good.  Thoughts in general on this?  What if it were a more commonly used method?

Reply:
Also, given that, why not just neuter the method entirely?

Reply:
> Ok, sounds good. Thoughts in general on this? What if it were a more commonly used method?

My general thought here is that; just from reading the docs (and completely ignoring the nullable annotations/feature); I would have no reason to believe I should need a null check here and would likely consume `Type.GetMethods()` without the null check.

Reply:
> What if it were a more commonly used method?

Case-by-case. If this null check was in a general purpose reflection helper that the code is likely going to pass a custom reflection objects into, I would keep it. Out of curiosity, I would also looked at history to see whether the null check was added as a bug fix, or whether it was just a defensive programming to start with. https://github.com/dotnet/coreclr/pull/20252 is an example of such superfluous null check that is worth keeping.

> why not just neuter the method entirely?

There may be somebody somewhere using this method for what it does...

Reply:
>  Out of curiosity, I would also looked at history to see whether the null check was added as a bug fix, or whether it was just a defensive programming to start with. 

It was included as part of the initial CoreCLR code population and I'm not quite sure the appropriate place to start digging deeper for the TFS sources that far back.

Reply:
Most of the history is in SourceDepot. I have checked this one and this null check was there in 2001 (2001 is as far as the history in SourceDepot goes).

Reply:
I'm fine with either keeping it or removing it. But, I feel that if we keep it we should likely update the documentation to indicate that someone could potentially return `null` here; as most developers are going to follow what the documentation says for things like `Type.GetMethods()`.

I also imagine that, regardless of `nullable annotations` not providing a "guarantee" that something won't return null (even if annotated that way); people are going to start relying on these annotations and removing checks from their own code if the API says it won't return `null` and logging bugs if it turns out the API is not annotated correctly.

Reply:
Keep the null check removed here and be done with it. It is not worth to spend any more time on this specific case.

`Type.GetMethods()` should be annotated as returning non-null, to match the documented contract.